src/pymor/reductors/dwr.py

```diff
@@ -0,0 +1,260 @@
+    dual_foms
+        List of the dual |Models| that correspond to each dimension of the output_functional.
+        If `dual_foms` is `None`, the default dual models are constructed
+        by :meth:`~pymor.reductors.dwr.create_dual_model`, assuming a fully discrete perspective.
+    primal_RB
+        |VectorArray| containing the reduced basis on which to project the fom.
+    dual_RBs
+        List of |VectorArrays| containing the reduced bases on which to project the `dual_foms`,
+        where each entry of the list corresponds to the dimensions of the output functional.
+        If `dual_bases` is `None`, the primal basis are used.
+    product
+        See :class:`~pymor.reductors.coercive.CoerciveRBReductor`.
+    coercivity_estimator
+        See :class:`~pymor.reductors.coercive.CoerciveRBReductor`.
+    check_orthonormality
+        See :class:`~pymor.reductors.basic.ProjectionBasedReductor`.
+    check_tol
+        See :class:`~pymor.reductors.basic.ProjectionBasedReductor`.
+    """
+
+    def __init__(self, fom, dual_foms=None, primal_RB=None, dual_RBs=None, product=None,
+                 coercivity_estimator=None, check_orthonormality=None, check_tol=None):
+        self.__auto_init(locals())
+        self._last_rom = None
+
+        if dual_RBs is not None:
+            assert len(dual_RBs) == fom.dim_output
+        assert (fom.output_functional is not None and fom.output_functional.linear), \
+            'DWRCoerciveRBReductor requires a linear ouput functional. ' + \
+            'Please use CoerciveRBReductor instead.'
+
+        self.primal_reductor = CoerciveRBReductor(fom, RB=primal_RB, product=product,
+                                                  coercivity_estimator=coercivity_estimator,
+                                                  check_orthonormality=check_orthonormality,
+                                                  check_tol=check_tol)
+        # construct default dual models if not provided
+        self.dual_foms = dual_foms or [self.create_dual_model(fom, d) for d in range(fom.dim_output)]
+        self.dual_reductors = []
+        # either needed for estimation or just for the corrected output
+        for d in range(fom.dim_output):
+            # construct dual model
+            dual_model = self.dual_foms[d]
+            # define dual reductors (with None as coercivity_estimator)
+            dual_basis = primal_RB if dual_RBs is None else dual_RBs[d]
+            dual_reductor = CoerciveRBReductor(dual_model, RB=dual_basis, product=product,
+                                               coercivity_estimator=None,
+                                               check_orthonormality=check_orthonormality,
+                                               check_tol=check_tol)
+            self.dual_reductors.append(dual_reductor)
+
+    def reduce(self, primal_dim=None, dual_dims=None):
+        primal_dim = primal_dim or len(self.primal_reductor.bases['RB'])
```

Comment:
This yields unexpected results for `primal_dim=0`.

Reply:
Very true. I'll fix it, thanks !